proposals/0000-integrations.md

```diff
@@ -0,0 +1,171 @@
+- **Libraries:** [`tailwind`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/tailwind/index.js), [`partytown`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/partytown/index.js), [`turbolinks`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/turbolinks/index.js)
+- **Features:** [`sitemap`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/sitemap/index.js)
+
+## Integration Usage API
+
+```js
+// astro.config.js
+export default ({
+	integrations: [
+		import('@astrojs/vue'),
+        // or, with options
+		[import('@astrojs/vue'), {/* options */}],
+        // or, as a static ESM import at the top of the file
+		vuePlugin({/* options */})
+	],
+});
+```
+
+## Integration Author API
+
+```ts
+export interface AstroIntegration {
+	name: string;
+	hooks: {
+		/** SETUP */
+		/** Called on astro startup, lets you modify config */
+		'astro:config:setup': (options: ConfigSetupOptions) => void | Promise<void>;
+		/** Called after config is finalized, lets you store config object for later */
+		'astro:config:done': (options: { config: Readonly<AstroConfig> }) => void | Promise<void>;
+
+		/** DEV */
+		/** Called on server setup, lets you modify the server */
+		'astro:server:setup': (options: { server: vite.ViteDevServer }) => void | Promise<void>;
+		/** Called on server startup, lets you read the dev server address/URL */
+		'astro:server:start': (options: { address: AddressInfo }) => void | Promise<void>;
+		/** Called on server exit */
+		'astro:server:done': () => void | Promise<void>;
+
+		/** BUILD */
+		/** Called on build start, lets you modify the server */
+		'astro:build:start': () => void | Promise<void>;
+		/** Called on build done, lets you read metadata about the build */
+		'astro:build:done': (options: { pages: string[]; dir: URL }) => void | Promise<void>;
+	};
+}
+```
+
+
+### Integration Author API: Hooks
+
+- The **Hook** is the main primitive of this proposed integration system.
+- Hooks optimize for maximum flexibility for the integration author: you can use our provided helper methods to perform common tasks during each hook, or write your own custom logic for advanced needs.
+- Hooks are conceptually aligned with how Rollup and Vite plugins work. This lets us pass some hooks (like 'astro:server:start') to Vite (the Vite `configureServer()` hook) with trivial effort.
+- The `hooks: {}` API conceptually matches Rollup & Vite but was designed to avoid the risk of conflict that would have been introduced had we literally extending the Vite plugin idea. 
+	- This is why we prefix all hooks with `astro:`.
+
+
+# Drawbacks
+
+- **Breaking changes across an integration system are expensive.** This can be mitigated until v1.0.0, see adoption strategy below.
+
+
+# Alternatives
+
+A previous design used a functional API more like this:
```

Comment:
I want to call out that 11ty uses this system as well. It has one clear drawback: **asynchronous blocks are harder to support.** 11ty has some internal reasons why asynchronous config functions aren't possible. But even if async _was_ supported, users may accidentally block crucial, possibly unrelated code in the config. I think dividing Astro's lifecycle into hooks will help us manage promises more effectively!